cheatsheets_draft/GraphQL_Cheat_Sheet.md

```diff
@@ -0,0 +1,257 @@
+
+#### Introspection + GraphiQL
+
+Many implementations of GraphQL have Introspection and GraphiQL enabled by default and leave them accessible without requiring authentication. This is problematic because introspection allows the requester to learn all about supported schema and queries (see a [real-world example](https://hackerone.com/reports/291531) abusing this). Introspection might be how the API owner wants to educate consumers about how to use the API. However, the preferred way to educate consumers about a service is through a separate documentation channel such as a wiki, Git Readme, or readthedocs.
+
+The safest and usually easiest approach is to just disable introspection and GraphiQL system-wide. See [this page](https://lab.wallarm.com/why-and-how-to-disable-introspection-query-for-graphql-apis/) or consult your GraphQL implementation's documentation to learn how to disable introspection altogether. If your implementation does not natively support disabling introspection or if you would like to allow some consumers/roles to have this access you can build a filter in your service to only allow approved consumers to access the introspection system.
+
+> This blurb is from nikitastupin. Would be great if we knew the name of this "hint" feature and link to documentation to disable it. I couldn't find anything from a quick Google.
+
+Keep in mind that even if introspection is disabled you can still guess fields by brute forcing them. Furthermore, some GraphQL implementations give you a hint when a field name you provide is similar to an existing field (e.g. you provide `usr` and the response may ask if you meant the valid `user` field instead). You should consider disabling this feature.
+
+***Disable Introspection - Java***
+
+```Java
+GraphQLSchema schema = GraphQLSchema.newSchema()
+    .query(StarWarsSchema.queryType)
+    .fieldVisibility( NoIntrospectionGraphqlFieldVisibility.NO_INTROSPECTION_FIELD_VISIBILITY )
+    .build();
+```
+
+***Disable Introspection & GraphiQL - JavaScript***
+
+```javascript
+app.use('/graphql', graphqlHTTP({
+  schema: MySessionAwareGraphQLSchema,
++ validationRules: [NoIntrospection]
+  graphiql: process.env.NODE_ENV === 'development',
+}));
+```
+
+### IDOR Protection
+
+In general, proper access controls will prevent any IDOR attacks since IDOR merely represents 2 separate authorization issues: [Broken Object Level Authorization](https://github.com/OWASP/API-Security/blob/master/2019/en/src/0xa1-broken-object-level-authorization.md) and [Broken Function Level Authorization](https://github.com/OWASP/API-Security/blob/master/2019/en/src/0xa5-broken-function-level-authorization.md). See the [IDOR prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Insecure_Direct_Object_Reference_Prevention_Cheat_Sheet.html) for a comprehensive rundown of preventing IDOR.
```

Comment:
I am not sure whether this is clear enough, but I'll wait for others' feedback.

"Classic" IDOR is in fact both BOLA and BFLA, but considering the GraphQL scope I would say that BFLA is an access control issue specific to GraphQL mutations (functions).

May be we can add something about BFLA in the "Mutation Access (Data Manipulation)" section.

Reply:
I would argue that "IDOR merely represents 2 separate authorization issues".

When you have a case of BFLA: for example, an admin function, "export_all_users", that doesn't implement access control, so regular users can access it - I don't think anybody would call it IDOR.
On top of that, the protection against IDOR/BOLA, is very different than protection against BFLA.

Reply:
Yeah I agree @inonshk, BFLA should not be called out as IDOR. I added this section based on feedback without thinking too much about it. If a user can access a function they shouldn't have access to that just should not be called IDOR in my mind.

However I do think BOLA is a traditional IDOR. The whole reason I initially added IDOR to this CS is because GraphQL usually involves the caller providing object IDs directly to access them. But I am thinking I will remove this IDOR section. Then I will add a sub bullet point to the first bullet point under the "Access Control" section that says "This will prevent [IDOR](https://cheatsheetseries.owasp.org/cheatsheets/Insecure_Direct_Object_Reference_Prevention_Cheat_Sheet.html)/[BOLA](https://github.com/OWASP/API-Security/blob/master/2019/en/src/0xa1-broken-object-level-authorization.md) issues." I could also add a new sentence to the "Mutation Access (Data Manipulation)" section and mention that doing proper checks for access to mutation would prevent [BFLA](https://github.com/OWASP/API-Security/blob/master/2019/en/src/0xa5-broken-function-level-authorization.md), but I don't think it's necessary to add.

Thoughts on that?

Reply:
I'm gonna move forward with this change since I haven't heard anything.